addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusHandlerFactory.java

```diff
@@ -68,10 +80,12 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
         ThingTypeUID thingTypeUID = thing.getThingTypeUID();
         if (thingTypeUID.equals(THING_TYPE_MODBUS_TCP)) {
             logger.debug("createHandler Modbus tcp");
-            return new ModbusTcpThingHandler((Bridge) thing, () -> manager);
+            ModbusEndpointThingHandler handler = new ModbusTcpThingHandler((Bridge) thing, () -> manager);
+            return registerEndpointHandler(handler);
```

Comment:
Please use the thing handlers `getServices()` method to register services for specific thing handlers.
Maybe I just misunderstood, in that case a bit more javadoc would be helpful.

Reply:
Thank you, I like this new concept, because it could solve the above problem as well.
I digged deep into the deconz binding as an example, but encountered a problem:  after I've removed the `@Component(service = DiscoveryService.class)` annotation, my service does not get registered with the discovery service anymore.
@davidgraeff  can you please give me a hint how the `ThingDiscoveryService` in deconz package gets registered with the DiscoveryService?

Reply:
Have you "registered" your service via the `getServices` method (that I linked above)?
If so: I think the annotation should be left in. It is still an OSGi service, just not manually started by you, but by the framework for your thing handler.

Reply:
Yes, I've did that, and the Services gets instantiated all right. But I noticed that the service does not get registered because of https://github.com/eclipse/smarthome/blob/b8455de15d65512e8fac4e94d42de6ab2fccf1c5/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java#L168-L173
So opposed to https://github.com/openhab/openhab2-addons/blob/master/bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/discovery/ThingDiscoveryService.java I've subclassed the ThingHandler interface. Now my interface gets registered but the DiscoveryServiceRegistryImpl. I tried both keeping the annotation and also removing it.
@davidgraeff can you please tell me why does it work with deconz? I do not see any component registration as in https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html#registering-as-an-osgi-service 

Alternatively can you point me to a cleaner example?

Thank you very much

Reply:
It never really worked for us in deconz, and I assume that was the reason. Good examples are the former eclipse smarthome extensions (Hue, Ikea Tradfri)

Reply:
Ok, I've sorted this out.

Obviously this case is more complicated than adding an action class to a handler. I've created a service class `ModbusEndpointDiscoveryService` which attaches to the handler instance like the action classes. And then this is responsible for executing the discovery processes. 

But we also want to make the discovery process extensible as well, because actual implementations will go into separate bundles. So there are the `ModbusDiscoveryParticipant` who also should be identifiable.

For the two to meet in the middle there is the ModbuscDiscoveryService, which is a stand alone service class, that collects the references both to the `ModbusEndpointDiscoveryService` objects and to the `ModbusDiscoveryParticipant` objects, also it gets registered at the DiscoveryServiceRegistryImpl.

Then when a new discovery scan is required it will orchestrate the communication between the  two families.